### ISSUE TITLE
lost telnet & ftp upon upgrade to 2.1.0.0A_201703071456 firmware

### ISSUE THREAD

**Issue by andy2301** [xmflsct/yi-hack-1080p#5]
My 1080p v1 comes with firmware 2.0 (2.0.0.1A_201612051401). I was able to apply this hack to get telnet and ftp. Thanks to @xmflsct!

However, yesterday I upgraded (through the YiHome app) to the latest 2.1 (2.1.0.0A_201703071456) firmware, and the hack doesn't work any more. Yes i tried to reapply the hack from scratch. I lost telnet & ftp. 

Also tried to store the firmware using the home_y20m dump generated by @xmflsct as posted in issue #2, but it didn't restore at all (renamed it to "home" and put at root of sd card). I suspect it's a generated home partition dump, not an official firmware.

** I'd appreciate if anybody can share a copy of the official firmware 2.0.0.1A_201612051401 for 1080p v1.**

Btw, you can also download the latest firmware 2.1.0.0A_201703071456 here. 
http://www.xiaoyi.com/home/firmware/download/?version=4
Click on "download" of the "2.1.0.0A_201703071456" firmware

**Comment by xmflsct** [comment]
I did not even dare to upgrade myself. :D So I have no idea what has been changed in the newer upgrade.

But feel free to use [my copy](https://github.com/xmflsct/yi-hack-1080p/issues/2#issuecomment-291467480), or other copies provided in that thread.

**Comment by andy2301** [comment]
@xmflsct I tried to restore your copy through the "manual firmware upgrade" approach, but it didn't do anything. I'm still with the upgraded 2.1 version. Do you know another way to apply/use your copy? 

My guess is that the new firmware gets rid of the factory_test.sh hole in its init.sh file. I tried to look into the firmware through binwalk but it appears the firmware is encrypted (binwalk returns nothing).

**Comment by shadow-1** [comment]
@andy2301 
Yes you are absolutely correct. I have confirmed with the Yi Dome cameras that firmware upgrades remove /home/app/script/factory_test.sh (which in turn calls the script from microSD card). Hence why my own custom firmware goes in a different direction.

I can create a version of my custom firmware for this camera. However I can't test it as I don't have this camera. You MUST have access to the terminal through the serial port for troubleshooting.

As mentioned elsewhere. Official firmware upgrade file is not really a firmware file. It is an encrypted archive. Explanation and instructions on how to unpack below.
https://github.com/fritz-smh/yi-hack/issues/124#issuecomment-275670843

**Comment by andy2301** [comment]
Great explanation of the firmware! @shadow-1 
I'll get a serial port later tonight for troubleshooting.

**Comment by xmflsct** [comment]
Hello again guys. I tried to reflash my home partition with @shadow-1's copy and my copy, both doesn't work somehow. I remember I used it when I restored my camera last time when I bricked it.. Now I have detached the serial connection just today... Tomorrow I will attach them again and see what happened, and try to produce a new flashable image.

**Comment by xmflsct** [comment]
Oops, my copy is actually working again somehow. @andy2301 Do you still have it, and would you like to try again?
My yellow flashes slowly when flashing, in between it flashes very quick for a second then comes back. After flashing is done, I have access to my cam again luckily.

**Comment by andy2301** [comment]
How did you flash it? 

When I tried it this morning, I did the following: 
I downloaded the zip file you posted, extracted to the root of sd card, rename it to "home". But it did't do anything for me. Upon reboot, it's still my upgraded 2.1 version.

**Comment by shadow-1** [comment]
@andy2301 
Do not rename the file. It needs to be named home_y20.
In addition, make sure your microSD card is FAT32 formatted. By default high capacity SD cards (over 32gb) are formatted using exFAT.

**Comment by xmflsct** [comment]
@andy2301 Aha, that's why. Because both @shadow-1's and mine copy are for uboot to use, which means that it differs from the official way of updating it. So as said, you should keep it as home_y20.

Very interesting thing is that. With just @shadow-1's copy, it does not work (I have no serial output to examine the problem now). After using @shadow-1's copy, I use mine again which gives a bit different yellow led flashing pattern and that worked - at least to bring the brick device back.

**Comment by shadow-1** [comment]
@xmflsct 
U-Boot does a check to see whether the recovery image matches what is currently on the system. If it thinks they are the same, the firmware upgrade is skipped. Considering it is U-Boot that is doing the check, I believe the check must be quite primitive.

From experimentation (I have flashed my camera probably over 50 times debricking and getting the region ban workaround working), it appears that the check to see whether the firmware on the system matches the recovery image is based on the embedded name within the image and the timestamp. So most likely, your current recovery image will no longer work as it matches what is currently on the system. However my one will work now because (most likely) the name is different and the timestamp is definitely different.

**Comment by xmflsct** [comment]
@shadow-1 
Good one! I need more debricking to catch up with your experience. :D

So I assume U-Boot thinks your image should be flashed, though it (might?) contain some error. After then I use my long-ago image, U-Boot now believes my image should be flashed.
Anyway, my long-ago image was simply wrapped the dd output with an uImage header, not as smart as your solution yet. I still need to figure out how I can properly pack jffs2 images. I think your v3 approach should be the way to go, as XiaoYi is tightening rules against hacking.

**Comment by shadow-1** [comment]
@xmflsct 
My image is unlikely to contain an error (although it might). I created a fresh jffs2 filesystem with the files from your long-ago backup. You are restoring the entire partition including the free space whilst my one skips the free space. Hence my image is smaller and a little quicker to flash. The end result is exactly the same as we are restoring exactly the same files. The only difference is the uImage header will be slightly different.

**Comment by xmflsct** [comment]
@shadow-1 Yes, I fully understand this part. :) Will first follow your approach and update this project, then see how far I can go with RTSP support as this is the function that I need - I don't care if I can use it with app or not. I have another 720p one working right now as a security cam, and would like to have another one. :)

**Comment by shadow-1** [comment]
Tomorrow I can send you an experimental version of my v3 firmware for this camera with working workaround for the region ban. However it will involve flashing the rootfs partition. As mentioned in the other thread, you have to upgrade uClibc which is almost impossible to upgrade on a running system on these cameras. Even if it were possible, it is highly risky and it is likely to brick. Using the offline upgrade through U-Boot is a much safer option.

I am confident that you will be able to restore your rootfs partition in case the camera gets bricked. The recovery method is exactly the same as for the home partition.

**Comment by xmflsct** [comment]
Thank you so much in advance! I will have a try of your ones and let you know if it is working.

Saying that, I have a [040 version](https://www.dropbox.com/s/xjzlvytuot8asvz/Hi3518E-V200R001C01SPC040.rar?dl=0) of SDK. I think they should have 050 by now. I will share it if I can find it and download it from somewhere.

**Comment by shadow-1** [comment]
Thanks for the updated SDK. I knew the 040 version has been released, however I didn't find a download link for it yet. When I got the 030 version, I don't think the 040 version was released (perhaps it was very new). I haven't heard anything about a 050 version yet.

With a little luck HiSilicon may have provided an example RTSP server with the SDK. Unlikely...but we can always be hopeful.

**Comment by xmflsct** [comment]
You are welcome. :) Not sure what had been updated though. Hope it would be a bit useful to you. 050 is not released yet, at least not on the forum where I downloaded 040. While browsing, I found a post with an updated sample code that claims to support RTSP for Hi3518EV200. I will have a try tomorrow, see if I can compile it correctly. Otherwise I might need your help again. :)

**Comment by andy2301** [comment]
@xmflsct Using your copy of home_y20, I've got my camera back to version 2.0.0.1A_201612051401. That's awesome!

Just extract it under the root directory of sd card and reboot. Make sure the name is "home_y20".

**Comment by xmflsct** [comment]
@andy2301 Glad that it works!

Do you happen to have serial connection as well? If so, can you try to change the boot delay and see if you enter U-Boot [fritz-smh#141 (comment)](https://github.com/fritz-smh/yi-hack/issues/141#issuecomment-283756414) My camera doesn't respond to any key press signal, so I was wondering whether they tweak the hardware or just a faulty part in my camera.

If you succeed in doing so, it would be nice if you could upgrade to the newest version again, and dump out the entire system in U-Boot. That provides an overview of what has been updated, and could be a base for the next step of this project. Let me know if this is feasible for you. :)

**Comment by shadow-1** [comment]
@xmflsct 
No need to tweak the U-Boot environment as you know how to flash rootfs and home partitions now. It should be a simple change on the env partition to get the countdown working if you really want to.

There is no need to upgrade the firmware to the newest version and dump the entire filesystem. I have provided instructions on how to extract the firmware and the upgrade is simply a replacement of the files on the camera within the /home directory.

Even so, if you want to make a backup of the camera and have serial access. It is easiest to use dd once booted rather than mucking around with U-Boot.

**Comment by xmflsct** [comment]
@shadow-1 
That's true. Then I will wait for your updated release, and combine it into this 1080p camera. :)

**Comment by shadow-1** [comment]
@xmflsct 
I got around to creating a test firmware for the Yi 1080 Home. The links are below:
[rootfs_y20](https://www.dropbox.com/s/si56hg7jx32byvz/rootfs_y20?dl=0)
[home_y20](https://www.dropbox.com/s/ms4rwovxeann00d/home_y20?dl=0)

Creating the home partition took longer than what I thought it would. I need to create some more build scripts to make it fully automated.

To make the upgrade more controlled and to make it easier for issues to be identified and rectified, what I recommend to do is follow the following steps:
1. Upgrade *only* the rootfs partition.
2. Boot up the camera and check that everything still works just as before (telnet working etc.) If there is an issue at this point, we will know that I deleted something I wasn't meant to and it's preventing bootup.
3. Flash the home partition.
4. Test whether telnet is working, FTP is working, web server is working (all services are enabled by default on the firmware).
5. Enter a valid proxychains configuration with at least one proxy server from Mainland China. This will allow your camera to not complain about the region ban and be fully operational with the Xiaomi smartphone app.

**Comment by xmflsct** [comment]
@shadow-1 
Thank you so much! I will try it just now and let you know how it goes.

By the way, the code I mentioned yesterday, I tried to compile it again, and still get the annoying `line 1: syntax error: unexpected word (expecting ")")`. All posts on the internet suggest that I might output an object file, but it is actually not - I test it with `file` which shows it as an executable. I suspect there might be some wrong flags in SDK's `Makefile.param` but my knowledge is not enough to find it out.
Would you like to have a try? Here is the archive.
[hp-ipc_rtsp.zip](https://github.com/xmflsct/yi-hack-1080p/files/896370/hp-ipc_rtsp.zip)

**Comment by maysamsh** [comment]
@shadow-1 I tried the rootfs, it revived the cam! When I tried to pair it with the phone it told me "This camera can only used within China", then I tried the other fine, home_y20, but unfortunately it didn't work and it's once again the status is still yellow.

**Comment by shadow-1** [comment]
@maysamsh 
Congratulations, it looks like your problem was with the rootfs partition afterall and it has now been restored.

This is completely experimental firmware and I have done quite a lot of changes to the home partition. I may have made a mistake with it because of all the changes I made without testing. You can restore the original home partition. You can use @xmflsct copy or my one and your camera will be back to how it was.

**Comment by xmflsct** [comment]
@maysamsh Yes, with my own copy of home_y20, I get my cam back and I can use it as before.

@shadow-1 Same happened to me. Flashing rootfs is fine, but bricked after home_y20. Still, I cannot run the proxychains4 from your home_h20 extract. I copy both the executable and the .so to exact the same location on my cam, I get `Illegal instruction`. The `ld-uClibc-0.9.33.2.so` in my /lib seems to be updated already with a last modify date at 03/06/17. Is that correct?

**Comment by maysamsh** [comment]
@shadow-1 Unfortunately it does not work anymore, even I have tried restoring images you already provided. What was the magic?!

**Comment by shadow-1** [comment]
@xmflsct 
Ah ok...looks like I tried to do too much at once.

I manually updated the firmware to 2.1.0.0A_201703071456.
I must have messed it up (I did it quite quickly), I'll redo to firmware upgrade with 2.0.0.1A_201612051401 firmware which we know works properly.

Yes I updated ld-uClibc-0.9.33.2.so already from a previous compile. You will also notice that Busybox on rootfs has also updated to include a few extra but useful utilities.

**Comment by shadow-1** [comment]
@xmflsct 
I am very confused why you get Illegal instruction error with those programs.
Can you provide me with the output of:
`cat /proc/cpuinfo`

**Comment by maysamsh** [comment]
@shadow-1 How could you manage to update to 2.1.0.0A_201703071456? When I put it into the SD and then to the camera, it does nothing.

**Comment by xmflsct** [comment]
@shadow-1 Yes, the rootfs is successfully updated, and I will just leave it there. I quickly examine the test home_y20 you provided. It seems like the /lib inside is left empty? For the rest files, I don't see why the booting stucks. If you think serial output would help, I can solder again.

`cat /proc/cpuinfo` writes:
```
Processor       : ARM926EJ-S rev 5 (v5l)
BogoMIPS        : 269.10
Features        : swp half fastmult edsp java
CPU implementer : 0x41
CPU architecture: 5TEJ
CPU variant     : 0x0
CPU part        : 0x926
CPU revision    : 5

Hardware        : hi3518ev200
Revision        : 0000
Serial          : 0000000000000000
```
I am also very confused of either illegal instruction or line 1 error...

**Comment by shadow-1** [comment]
@xmflsct 
In the rush of creating this firmware...i copied the files from your original firmware file incorrectly and I missed many files. It had no hope of working as the lib folder was empty. For some reason your home partition backup has a program called tempbox and a file named index.html.

Did this come with the official firmware, or is it something you were trying out on the camera before you made your backup?

The processor on the Yi 1080p Home and the Yi 1080p Dome are exactly the same. Everything I compile and tested on my camera should work on your camera...

**Comment by shadow-1** [comment]
@maysamsh 
I believe you will need to use @xmflsct backup of the home partition for your camera to get back operational. Hopefully you didn't go messing around with various backups of rootfs partition.

**Comment by xmflsct** [comment]
@shadow-1 
That tempbox was my very early trail of busybox, that is a pre-compiled one, and I just rename it into that... You can delete that tempbox for sure. And the index.html was there in the original firmware, and I had absolutely no idea why it would be useful at all.

So... Here comes the trick... I didn't use any rsync or similar program to upload binary files. And my FTP program somehow did not upload binary files as it is... Now it is solved, and I can run proxychains4! I will try other programs that failed before, and maybe they are fine now... My bad!

**Comment by maysamsh** [comment]
@shadow-1 seems it's screwed up. I'm going to get the serial connection!

**Comment by xmflsct** [comment]
@maysamsh That's a good attitude. :) I can feel that with @shadow-1's very generous help, we are almost there!

**Comment by shadow-1** [comment]
@xmflsct 
Try this home partition:
[home_y20](https://www.dropbox.com/s/ms4rwovxeann00d/home_y20?dl=0)

I still manually upgraded the firmware to 2.1.0.0A_201703071456.
Hopefully now I didn't miss copying whole folders!

**Comment by xmflsct** [comment]
@shadow-1 It works!

**Comment by shadow-1** [comment]
@xmflsct 
Fantastic news!
I find the web interface quite handy.

Once you enter some valid proxy servers from Mainland China into the ProxyChains configuration on the web interface, the camera will work with the smartphone app.

**Comment by maysamsh** [comment]
@shadow-1 It works! but still I have the region ban issue, i guess, after trying to pair the status LED becomes still blue.

**Comment by xmflsct** [comment]
@shadow-1 @maysamsh Yes, I have the same region ban issue as well. I have tried several proxies, and none works. Those proxies are confirmed working on computer.

Strange thing is, if I use `proxychains4 ping google.com` or `proxychains4 ping facebook.com`, it should not get through.. But both I get good pings. How can I confirm proxychains4 is working correctly?

**Comment by andy2301** [comment]
Excellent! It looks you guys are making great progress.

@xmflsct Try the following. It'll show your WAN ip without and with proxy.

wget http://ipinfo.io/ip -O -
./proxychains4 -f ./proxychains.conf wget http://ipinfo.io/ip -O -

**Comment by xmflsct** [comment]
@andy2301 Thanks! Found some free socks proxies online, wget works fine, but still I cannot connect to my cam using app. Anyway, I need to figure out RTSP first as that's what I really need.

**Comment by andy2301** [comment]
@xmflsct The proxychains-ng approach works with the region ban.

1. make sure proxychains4 and libproxychains4.so are under /home/app
2. make sure /etc/proxychains.conf has the correct proxy configuration. Remember to test your proxy settings elsewhere and make sure it works.
3. run the following (or make sure it's like this)
    mv /home/app/cloudAPI /home/app/cloudAPI_real
    echo "/home/app/proxychains4 /home/app/cloudAPI_real \"$@\"" > /home/app/cloudAPI
    chmod 755 /home/app/cloudAPI
4. reboot your camera. You should be able to get rid of region ban, plus with YiHome app access.

Thank you @shadow-1 for the help and great instructions!

**Comment by andy2301** [comment]
@xmflsct to reply to you earlier question in [your post](https://github.com/xmflsct/yi-hack-1080p/issues/5#issuecomment-291804451) , I didn't use my serial port cable, since I got telnet access after restoring your home_y20 image successfully.

**Comment by xmflsct** [comment]
@andy2301 No worries at all! I am still trying to find a working proxy to test everything. Some proxies work on PC, but get a deny when I use it with proxychains4.

Besides, would you like to take a look at [hp-ipc](https://github.com/xmflsct/yi-hack-1080p/files/896370/hp-ipc_rtsp.zip) as well?

**Comment by andy2301** [comment]
@xmflsct also, i just tried your hp-ipc_rtsp.zip. It compiled and it ran but caused the camera to reboot in the end.

```
/tmp/sd # ./hp_ipc 
linear mode
Open /dev/spidev0.0 error!
SPI_IOC_MESSAGE error 
SPI_IOC_MESSAGE error 
<repeat .....>
SPI_IOC_MESSAGE error 
-------Sony IMX222 Sensor Initial OK!-------
RTSP:-----Init Rtsp server
u32Stride: 3840
===============u32Size: 4147200
VbBlk: 0x20000

 please turn off camera aperture!

hit the Enter key ,start calibrate!
level = 0
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread

HI_MPI_ISP_FPNCalibrate err: 0xa0108007
FPN calibrate failed 0xa0108007!
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
Connection closed by foreign host.
```

**Comment by xmflsct** [comment]
@andy2301 It seems like you haven't update the `SENSOR_TYPE` in the `Makefile.param` in the sample folder. I read somewhere that this cam has OV9712 sensor, but somehow the SDK shows that it is a 720p sensor.. After I change the `SENSOR_TYPE` and run hp_ipc, I got:
```
linear mode
Not support! Width:1919, Height:1079, Fps:25.000000, WDRMode:0
[Func]:HI_MPI_ISP_Init [Line]:284 [Info]:ISP[0] set sensor image mode failed!
s32IspInit: HI_MPI_ISP_Init failed!
s32StartVi: Sensor init failed!
start vi failed!
```
It seems we might need to modify a bit more stuff. [Source code origin](http://www.ebaina.com/bbs/forum.php?mod=viewthread&tid=10950&extra=&page=1).

**Comment by xmflsct** [comment]
@andy2301 In the 040 SDK I shared, I found a very useful tool `PQTool` which might help us break through setting up hp_ipc. I can run the tool between my cam and PC, and grab still RAW images. The .ini file contains tons of useful info for setting up hp_ipc. But I am unfamiliar with C so I can be slow...

**Comment by shadow-1** [comment]
@xmflsct @andy2301 
The firmware I posted has incorporated all the necessary changes. `cloudAPI `application is moved to a different name, a script replaces `cloudAPI`, `proxychains.conf` is in the correct location `/home/app`, `proxychains4` is in the correct location `/home/app/localbin`, `libproxychains4.so` is in the correct location `/home/app/locallib`.

The whole trick is to find a reliable proxy server. I deliberately didn't provide any in the firmware as they tend to move around and in addition, if one gets too popular, it's likely it will get shutdown. I have managed to find reliable proxy servers.

The way I test the proxy server is to set it up as my browsers proxy server to the socks server I want to test and visit https://whatismyipaddress.com/

I forgot to mention a few requirements:
* The socks server must reside within Mainland China. Many free servers are located in Taiwan, Hong Kong, Thailand despite being described as China.
* The socks server must support HTTPS. Hence why I visit the https://whatismyipaddress.com/ and not http://whatismyipaddress.com/ Many support HTTP but nit HTTPS.

I will develop a test proxy server page on my web interface on the camera itself in a future update. I already know how to do it. However I need to spend time to write it and to make it refined.

**Comment by xmflsct** [comment]
@shadow-1 Thank you for the update! Indeed all your files are in the correct place. The trick is to find a socks server that support HTTPS.. This is where I failed most probably. I will keep looking.

**Comment by andy2301** [comment]
Sounds like good progress. Maybe 1919 is expected because the pixel sequence starts from 0 and extends to 1919 (1920 pixels in total). I'll take a look later tonight.

**Comment by xmflsct** [comment]
My last update today.
After fixing the pixels manually to 1280 x 720 in the source file, the program then runs but with following error.
```
linear mode
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
=========================================================
===ominivision ov9712 sensor 720P30fps(Parallel port) init success!=====
=========================================================
RTSP:-----Init Rtsp server
u32Stride: 2560
===============u32Size: 1843200
VbBlk: 0x20000

 please turn off camera aperture!

hit the Enter key ,start calibrate!
level = 0
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
```
This might trigger camera reboot, so before running it, you can try to kill those default apps initiated by the system, especially watchdog.

**Comment by andy2301** [comment]
@xmflsct I got the same error with hp_ipc but no further progress. Google search of this "venc stream time out" error shows that many people had something working but when they changed the sensor_type on a different board they would face the same error "get venc stream time out, exit thread".  So I guess we need to do the following:

1. Figure out what is the correct CHIP_ID and the SENSOR_TYPE to set in Makefile.param. 

Note that whatever sensor library the working p2p_tnp process uses must be present in the firmware. I noticed that the current firmware has ov9732 library (/home/app/locallib) but no ov9712 lib, so I tried the sensor_type of OV9732. But still the same error. 

I also tried a different CHIP_ID=CHIP_HI3518E_V200 but the same error. Is there a way to confirm absolutely what is the right value for these 2 parameters?

```
#CHIP_ID ?= CHIP_HI3516C_V200
#CHIP_ID ?= CHIP_HI3518E_V200
#CHIP_ID ?= CHIP_HI3518E_V201
#SENSOR_TYPE ?= OMNIVISION_OV9732_DC_720P_30FPS
```

```
/tmp/sd # ls /home/app/locallib/
libsns_f02.so     libsns_gc2023.so  libsns_ov9732.so
libsns_f22.so     libsns_ov2710.so  luaext.so
```

2. I feel that we need to make sure the basic mpp sample program compiles and works. 
Right now I have ittb_control + PQ_Tool works to capture an image remotely. However, ittb_control is just binary without source code. At the same time, I compiled and tried all of the vio or venc sample programs but none worked. That's the first gap we need to address, if we want to compile a rtsp server. If our compiled sample_venc program can't even dump a static image, there must be sth wrong with our compilation settings.

```
/tmp/sd/sample_venc # ./sample_venc
Usage : ./sample_venc <index>
index:
         0) 1*1080p H264 + 1*VGA H264.
         1) 1*1080p MJPEG encode + 1*1080p jpeg.
         2) low delay encode(only vi-vpss online).
         3) roi background framerate.
         4) Thumbnail of 1*1080p jpeg.
         5) svc-t H264
```

```
/tmp/sd/sample_venc # ./sample_venc 4
=============SAMPLE_COMM_VI_SetMipiAttr enWDRMode: 0
linear mode
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
I2C_WRITE error!
=========================================================
===ominivision ov9732 sensor 720P30fps(Parallel port) init success!=====
=========================================================
press 'q' to exit sample!
peress ENTER to capture one picture to file
********************* Open pwm device error! *********************

press the Enter key to snap one pic

[SAMPLE_COMM_VENC_SnapProcess]-1262: snap time out!
SAMPLE_VENC_1080P_JPEG_Thumb: sanp process failed!
please press ENTER to exit this sample

program exit abnormally!
```

**Comment by xmflsct** [comment]
@shadow-1 I can confirm that, with a working socks5 proxy with SSL support, I can pair and connect to my cam, with great streaming speed. :)

**Comment by maysamsh** [comment]
@xmflsct @shadow-1 Could you manage to hack it to bypass region ban? Something other than using a proxy.

**Comment by shadow-1** [comment]
@xmflsct Socks4 proxies with SSL support work perfectly as well. Haven't tested it with a HTTP proxy.

@maysamsh Sorry, did not find other methods to defeat region ban yet. Your camera must authenticate to the Xiaomi server. Xiaomi server makes the decision whether to region ban you, not the camera itself. If you skip the authentication, the camera will not work with the smart phone app.

**Comment by maysamsh** [comment]
@shadow-1 So basically it's useless without using a proxy 😞

**Comment by xmflsct** [comment]
@maysamsh That's why I am trying to figure out RTSP possibility.

**Comment by shadow-1** [comment]
RTSP is probably the best long term solution. However the Xiaomi smart phone app is very intuitive and works very well. Especially for those without a device that can act as a DVR server.

**Comment by xmflsct** [comment]
@andy2301 After playing around, I am still unsure whether the image sensor is really OV9712 or OV9732 or even others. The reasons are:
1. In `load3518e`, the sensor type is actually hardcoded.
2. When I started `HiIspTool.sh`, whether I use config of OV9712 or OV9732, I still can capture raw image in 1080p without problem. So I have no idea what is the trick with my knowledge.
3. When I rebuild OV9712 library into either static or shared one, after running `sample_venc`, the output of `cat /proc/umap/vi` changes from 1080p to 720p.
4. Other competing products in China mostly use OV2710 which is a truly 1080p sensor, but it is not part of the SDK from HiSilicon.

Assuming that the image sensor is actually a variation of OV9712, we might need to modify quite some addresses in `ov9712_sensor_ctl.c`. Without proper SDK of this sensor, it seems almost impossible to do so.

For now, I give up this path, unless you would like to try a bit more. I will see if I can find back any source code from MP77V's rtspsvr.

**Comment by shadow-1** [comment]
I can confirm that the processor is Hi3518ev200.
To verify, run the following command: `cat /proc/cpuinfo`

The bootup messages on my camera whilst the kernel is loading shows the following line:
`==== Your input Sensor type is ov9712 ====`

So I believe the correct sensor is ov9712.
You don't get to see the bootup messages when you login through telnet.

**Comment by xmflsct** [comment]
Yes, processor model is very certain. But the image sensor model is not.
If you take a look at `/home/hisiko/load3518e`, line 16 or nearby, the sensor type is hardcoded. And the `==== Your input Sensor type is ov9712 ====` just reflects this 'variable'.
The corresponding `load3518` file in my 720p version also has ov9712 hardcoded. That's why I have doubt in either the hardware or the software.

**Comment by xmflsct** [comment]
By deleting image sensor libraries in `/home/app/locallib/`, it is verified that none of the OV libraries is the correct one. Only `libsns_f22.so` will turn the cam into continuous reboot - after WiFi is set up, and when those customized software start to run which gives me few seconds to squeeze back `libsns_f22.so` and I don't need to reflash the home partition. :)

A quick look at this binary, its structure is similar to those OV ones, with `F22sensor 1080P25fps(Parallel port)`. So it is safe to assume this is actually the type of image sensor on board rather than OV9712.

**Comment by andy2301** [comment]
Excellent! So we should copy the f22 lib for cross-compiling and see if that makes a difference. I will also take a look tonight.

**Comment by andy2301** [comment]
Yesterday I also found that the "strings rmm" output has "DEV_ATTR_F22_DC_1080P_BASE", which seems to indicate the sensor type.

**Comment by andy2301** [comment]
N

**Comment by xmflsct** [comment]
With our two sources of info, I think it is certain that the image sensor is F22. I will now copy over that library, and see what compiles out.

**Comment by xmflsct** [comment]
The `I2C_WRITE error!`is gone! But somehow `sample_venc` always try to access frame rate of 30, but the sensor is not supported. I will see how I can reduce that. Also, every time when I run the program, the camera will reboot in some seconds. Don't know if it is done by the watchdog.

**Comment by andy2301** [comment]
Kill the watch process and it won't reboot.

**Comment by shadow-1** [comment]
You need to determine whether the actual watchdog on the board is rebooting the camera or the Xiaomi programs. I assume this implementation of RTSP is interfering with the Xiaomi applications which are already running. You can stop the Xiaomi applications from restarting or rebooting the camera automatically by killing the Xiaomi watch process:
`killall watch_process`

When you kill this, if there is a problem with a Xiaomi app, the system it will not attempt to recover from it.

If the actual watchdog on the HiSilicon chip is being triggered. It indicates a much more serious problem as the processor is detecting that the system has hanged (kernel panic or similar issue). Most hardware watchdogs are configured to automatically reboot.

**Comment by xmflsct** [comment]
It is Xiaomi's watchdog reboots the camera. My assumption is that, due to this implementation of RTSP (or currently the `sample_venc`) tries to initiate the sensor as if it were not there. This would break Xiaomi's applications which triggers Xiaomi's watchdog. I have killed all their program, and now it is not rebooting anymore. I still need to figure out if it is okay to re-initiate image sensor. At least kernel does not panic at this moment.

By the way, I cannot figure out in `sample_venc` where it tries to set frame rate to 30. So I would just move on to the hp_ipc example and see if it works.

**Comment by xmflsct** [comment]
Somehow in `hp_ipc.c`, the default frame rate is 25, and the sensor still complaints that it is not supported. After I reduce it to 15, the complaint is gone!

Now it seems that sensor is successfully initiated, at least no other complaints, and `cat /proc/umap/vi` looks fine. Now back to the original problem that I need to search for...
```
linear mode
sesnorID:0x80; sensoraddr:0x1; sensordata:0x1
=========================================================================
===F22sensor 1080P25fps(Parallel port) init success
=========================================================================
RTSP:-----Init Rtsp server
u32Stride: 3840
===============u32Size: 4147200
VbBlk: 0x20000

 please turn off camera aperture!

hit the Enter key ,start calibrate!
level = 0
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
```

**Comment by andy2301** [comment]
Compiled with libsns_f22.so, sample_venc works! I got a static image snap, and got the stream_chn0 file populated locally.

```
/tmp/sd/sample_venc # ./sample_venc 1
=============SAMPLE_COMM_VI_SetMipiAttr enWDRMode: 0
linear mode
sesnorID:0x80; sensoraddr:0x1; sensordata:0x1
=========================================================================
===F22sensor 1080P25fps(Parallel port) init success
=========================================================================
press 'q' to exit sample!
peress ENTER to capture one picture to file
********************* Open pwm device error! *********************
press the Enter key to snap one pic
snap 0 success!
```

We definitely made progress with hp_ipc. As you can see, not only the I2C_WRITE error is gone, but also we get messages of accepting RTSP connection. The problem is that, although hp_ipc shows it's working, my VLC client had no video showing up (it's not disconnecting through).

```
/tmp/sd/hp_ipc # ./hp_ipc_HD1080_25.f22 
linear mode
sesnorID:0x80; sensoraddr:0x1; sensordata:0x1
=========================================================================
===F22sensor 1080P25fps(Parallel port) init success
=========================================================================
********************* Open pwm device error! *********************
RTSP:-----Init Rtsp server
u32Stride: 3840
===============u32Size: 4147200
VbBlk: 0x20000

 please turn off camera aperture!

hit the Enter key ,start calibrate!
level = 0
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
<<<<RTSP Client 192.168.101.139 Connected...
RTSP:-----Create Client 192.168.101.139
<<<<<OPTIONS rtsp://192.168.101.81:554/stream_chn0.h264 RTSP/1.0
CSeq: 2
User-Agent: LibVLC/2.2.4 (LIVE555 Streaming Media v2016.02.22)


>>>>>RTSP/1.0 200 OK
CSeq: 2
Date: Thu, Apr 06 2017 21:54:32 GMT
Public: OPTIONS,DESCRIBE,SETUP,PLAY,PAUSE,TEARDOWN


<<<<<DESCRIBE rtsp://192.168.101.81:554/stream_chn0.h264 RTSP/1.0
CSeq: 3
User-Agent: LibVLC/2.2.4 (LIVE555 Streaming Media v2016.02.22)
Accept: application/sdp


--------------------------------------------
>>>>>RTSP/1.0 200 OK
CSeq: 3
Date: Thu, Apr 06 2017 21:54:32 GMT
Content-Type: application/sdp
Content-length: 337
Content-Base: rtsp:///stream_chn0.h264/

v=0
o=StreamingServer 3331435948 1116907222000 IN IP4 
s=H.264
c=IN IP4 0.0.0.0
t=0 0
a=control:*
m=video 0 RTP/AVP 96
a=control:trackID=0
a=rtpmap:96 H264/90000
a=fmtp:96 packetization-mode=1; sprop-parameter-sets=AAABBCCC
m=audio 0 RTP/AVP 97
a=control:trackID=1
a=rtpmap:97 G726-32/8000/1
a=fmtp:97 packetization-mode=1

<<<<<SETUP rtsp:///stream_chn0.h264/trackID=0 RTSP/1.0
CSeq: 4
User-Agent: LibVLC/2.2.4 (LIVE555 Streaming Media v2016.02.22)
Transport: RTP/AVP;unicast;client_port=65342-65343


--------------------------------------------
>>>>>RTSP/1.0 200 OK
CSeq: 4
Date: Thu, Apr 06 2017 21:54:32 GMT
Transport: RTP/AVP;unicast;destination=;client_port=65342-65343;server_port=53255-53511
Session: 1000

<<<<<SETUP rtsp:///stream_chn0.h264/trackID=1 RTSP/1.0
CSeq: 5
User-Agent: LibVLC/2.2.4 (LIVE555 Streaming Media v2016.02.22)
Transport: RTP/AVP;unicast;client_port=56898-56899
Session: 1000


--------------------------------------------
>>>>>RTSP/1.0 200 OK
CSeq: 5
Date: Thu, Apr 06 2017 21:54:32 GMT
Transport: RTP/AVP;unicast;destination=;client_port=56898-56899;server_port=53255-53511
Session: 1000

<<<<<PLAY rtsp:///stream_chn0.h264/ RTSP/1.0
CSeq: 6
User-Agent: LibVLC/2.2.4 (LIVE555 Streaming Media v2016.02.22)
Session: 1000
Range: npt=0.000-

;client_port=56898-56899
Session: 1000


--------------------------------------------
>>>>>RTSP/1.0 200 OK
CSeq: 6
Date: Thu, Apr 06 2017 21:54:32 GMT
Range: npt=0.000-
Session: 1000
RTP-Info: url=rtsp:///stream_chn0.h264;seq=0

udp up
Start Play
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
get venc stream time out, exit thread
<<<<<TEARDOWN rtsp:///stream_chn0.h264/ RTSP/1.0
CSeq: 7
User-Agent: LibVLC/2.2.4 (LIVE555 Streaming Media v2016.02.22)
Session: 1000

 npt=0.000-

;client_port=56898-56899
Session: 1000


>>>>>RTSP/1.0 200 OK
CSeq: 7
Date: Thu, Apr 06 2017 21:54:43 GMT
Session: 1000

RTSP:-----Exit Client 192.168.101.139
```

**Comment by xmflsct** [comment]
@andy2301 Wonderful! You have reached further! Somehow my sample_venc is still complaining about wrong FPS. Could be that I have messed too much these days in SDK. I will get a clean one and try again.

The problem with hp_ipc is that, as we read yesterday, there are quite some setting of setting `vi` that has to be correct. But if I can also get sample_venc running, I might be able to grab some info from there.
Anyway, if you can try to run `rmm` and `cat /proc/umap/vi`, compared with `hp_ipc` and `cat /proc/umap/vi` and see what's the difference. One obvious difference that I can easily modify is the mask. For the rest, my zero knowledge of handling image sensor won't help.

I was also thinking, what if let `rmm` initialize everything and we just _grab_ from venc. I have tried that, it is not possible. First hp_ipc will bark, then `rmm` will bark also with unintended request. So we have to really get setting vi right, before we can have proper RTSP stream.

**Comment by andy2301** [comment]
FYI. The following is the changes I made to Makefile.param, before compiling sample_venc and hp_ipc.
```
CHIP_ID ?= CHIP_HI3518E_V200
SENSOR_TYPE ?= OMNIVISION_OV9712_DC_720P_30FPS
....
ifeq ($(SENSOR_TYPE), OMNIVISION_OV9712_DC_720P_30FPS)
      #SENSOR_LIBS += $(REL_LIB)/libsns_ov9712.a
      SENSOR_LIBS += $(REL_LIB)/libsns_f22.so
endif
```

Yes I verified that cat /proc/umap/vi is correct with hp_ipc. vpss is correct too. I suspect the venc part of the code in hp_ipc is broken. Remember that sample_venc used to throw the same error "get venc stream time out, exit thread" but with the f22 lib it's gone.

I looked at the difference between these two functions:
```
SAMPLE_COMM_VENC_StartGetStream, SAMPLE_COMM_VENC_GetVencStreamProc in mpp/sample/common/sample_comm_venc.c (for sample_venc)
vs. 
s32StartGetStream, vdGetVencStreamProc in hp_ipc/src/hp_venc.c (for hp_ipc)
```

They both has the printf line of `get venc stream time out, exit thread`. There's some subtle difference in the coding. I suspect that sample_venc is more updated so it works with f22 lib. So a simple solution is to copy that code piece and apply it in hp_ipc. I don't have time to work on this for now. Will come back later tonight.

**Comment by xmflsct** [comment]
Thank you! Yes, that's what I modified as well, with a clean SDK 040 unpack, but still no luck. It is either complaint not supported frame rate of 25 or 30. This complaint comes from libsns_f22.so.
```
/home/app/localbin # ./sample_venc 1
=============SAMPLE_COMM_VI_SetMipiAttr enWDRMode: 0
linear mode
sesnorID:0x80; sensoraddr:0x1; sensordata:0x1
=========================================================================
===F22sensor 1080P25fps(Parallel port) init success
=========================================================================
Not support Fps: 30.000000
press 'q' to exit sample!
peress ENTER to capture one picture to file
Not support Fps: 30.000000
********************* Open pwm device error! *********************
^Cprogram termination abnormally!
```

Will leave it for now and go to sleep. Hope for your good news tomorrow. :) I will finish reading rest of the posts and see what I can learn from them.

**Comment by xmflsct** [comment]
FYI, I accidently found out that, one and a half hours ago, I did capture an image succesfully as `.mjp`, probably with `./sample_venc 4`. It means that the program used to run once somehow... Anyway...

**Comment by xmflsct** [comment]
@andy2301 Play around a bit more today, but getting no where. I tried to match those two functions as you suggested, they are indeed almost the same, but still get the same error. Then I look at my vi, vpss and venc. For my `cat /proc/umap/vi`, I still see `VI-VPSS is offline`. So obviously there is nothing in my vpss. In my venc, I can see at least an attempt to create one channel but failed. With `rmm`, there will be four channels created. My assumption is still, although there is no complaint during sensor initiation, I suspect there is something wrong there, which leads to no output to venc, as also [suggested by the original author](http://www.ebaina.com/bbs/forum.php?mod=viewthread&tid=10950&extra=page%3D1&page=1).

@shadow-1 @andy2301 
By the way, do you guys have any idea of these two files? `/tmp/view` and `/tmp/mmap.info`. The RTSP server provided in 720p hack was using these two files directly, and these two files are (seems to be) generated by `rmm`. My assumption is that, `mmap.info` provides essential mapping to memory, where `view` is the image data file in some sort of format. I have tried to open them in common image or video formats but without success. If we could make use of these two files, we can then bypass the difficult hardware bit, and be able to run (hopefully) RTSP server and official software at the same time.

**Comment by shadow-1** [comment]
@xmflsct 
I know what `/tmp/view' is. It is a h264 video buffer. Half the file is dedicated for the HD stream whilst the other half is dedicated for the SD stream. My original idea to get RTSP working on the camera was to stream directly from `/tmp/view`. That way, we can have the camera operational from both the smartphone app and from RTSP at the same time without putting much additional load on the camera.

My current assumption is that the one of the Xiaomi applications running on the camera grabs the live video and audio streams. A bit of post processing occurs on the live video and this post processed video is buffered to `/tmp/view` before streaming through their cloud application (which is a separate program running on the camera).

The main drawback of streaming directly from `/tmp/view` is that it is the post processed video feed. It contains the 'Yi' watermark in the bottom left corner as well and the date/time in the top left corner. This is exactly as what is seen in the smartphone app. However on the plus side, I believe the video feed is also improved through the post processing (lighting/sharpness adjustments).

I have successfully streamed `/tmp/view` using the live555 library. However I could not make it work on a continuously updating file (it is a live stream after all). My plan was after making streaming work on the continuously updating file was to extend the functionality so that that two streams are opened from the same file (remember portion of `/tmp/view` is the HD stream whilst portion of `/tmp/view` is the SD stream. I am currently on vacation and won't have time to develop RTSP at the moment.

I'm not sure what `/tmp/mmap.info` file is for. However I am familiar with all the strings that it contains. It appears to contain all the information that is sent to the Xiaomi server. When I was hacking the camera to defeat the region ban, I examined all the communications between the camera and the Xiaomi server and I can confirm that all the information contained in this file is being sent to Xiaomi server. Most of the information is accessible via the smartphone app. It is how the smartphone app knows the SSID of your network, wifi strength, ip address of the camera, mac address of the camera etc.

**Comment by andy2301** [comment]
@xmflsct Guess what, I have a working rtsp server! After quite a lot of debugging, I finally make it work, but with a couple of caveats.

1. it only streams 720P, not 1080P. Tried to make 1080P work but encountered too many errors.
2. it doesn't run at the same time with rmm. Running rtsp2301 will make rmm quit, and your camera will reboot. So remember to kill the watch_process before running rtsp2301. I'm attaching the rtsp2301 binary, feel free to give it a try.
[rtsp2301_binary.zip](https://github.com/xmflsct/yi-hack-1080p/files/907347/rtsp2301_binary.zip)



```
/tmp/sd/rtsp # ./rtsp2301
size of 2RTSP server START
Streaming URL: rtsp://192.168.101.81:554/ch0_0.h264
listen for client connecting...
enSize=16, u32BlkSize=1491840
=============SAMPLE_COMM_VI_SetMipiAttr enWDRMode: 0
linear mode
sesnorID:0x80; sensoraddr:0x1; sensordata:0x1
=========================================================================
===F22sensor 1080P25fps(Parallel port) init success
=========================================================================
         c) cbr.
please press twice ENTER to exit this sample
********************* Open pwm device error! *********************
Incoming RTSP connection accepted on socket: 16
EventLoop Connection reached: 3
192.168.101.139 Port:57299
```

I can live with 720P. But I definitely want rtsp2301 to co-exist with rmm. So this is still not a perfect solution.

@shadow-1 Now I understand much better about the RTSP process, I wonder if it's possible to stream directly from /tmp/view. Do you happen to know more about the structure of /tmp/view?

**Comment by xmflsct** [comment]
@andy2301 Wonderful news! At least you are reaching somewhere!
Sadly it won't run on my cam, with the exact same error that stopped me from progressing. It is still barking not support fps. Even I flash back the firmware provided by @shadow-1, still the same issue.
```
/home/app/localbin # ./rtsp2301
size of 2RTSP server START
Streaming URL: rtsp://192.168.1.81:554/ch0_0.h264
listen for client connecting...
enSize=16, u32BlkSize=1491840
=============SAMPLE_COMM_VI_SetMipiAttr enWDRMode: 0
linear mode
sesnorID:0x80; sensoraddr:0x1; sensordata:0x1
=========================================================================
===F22sensor 1080P25fps(Parallel port) init success
=========================================================================
Not support Fps: 30.000000
         c) cbr.
please press twice ENTER to exit this sample
[SAMPLE_COMM_VENC_GetVencStreamProc_RTSP]-75: HI_MPI_VENC_GetChnAttr chn[0] failed with 0xffffffff!
Not support Fps: 30.000000
********************* Open pwm device error! *********************
Config ISP register Failed with ec 0xffffffff!
Config ISP register Failed with ec 0xffffffff!
Config ISP register Failed with ec 0xffffffff!
```

**Comment by xmflsct** [comment]
Interestingly, when I run your program after I flash back to the old (before updated) rootfs and home, the program won't bark about wrong fps anymore! And RTSP stream works great, with some delay but acceptable! After then, I flash to @shadow-1's rootfs, still runs fine. Lastly, I flash to @shadow-1's home, the program barks again. Did you happen to modify any other file/library by any chance?

I copy the old `libsns_f22.so` to cam, the fps barking is gone.  But still I get `Config ISP register Failed with ec 0xffffffff!`. Are you sure you have successfully updated to @shadow-1's home build?

I will anyway try to track down which library/module that causes this problem.

**Comment by shadow-1** [comment]
@xmflsct 
I believe @andy2301 is not using my home partition and is using the original 2.0.0.1A_201612051401 firmware.

The home partition I made is based on version 2.1.0.0A_201703071456 of the firmware. The update to 2.1.0.0A_201703071456 replaced files in  `/home/app` and `/home/hisiko`.

**Comment by xmflsct** [comment]
@shadow-1 That would probably be the case. Don't know what has changed inside `libsns_f22.so` and other files...

**Comment by shadow-1** [comment]
@xmflsct I'm not sure what has changed in the update files. I noticed libsns library is provided in the SDK for various sensors. However not libsns_f22.so.

**Comment by shadow-1** [comment]
@andy2301 
I somehow missed your question.

Yes it is possible to stream from `/tmp/view`. I have managed to stream it as a file so far but not as a continuous stream using live555 library. I'm not sure whether live555 library is the best option in this case but I gave it a go as it was the only free RTSP server I knew about at the time.

`/tmp/view` is a h264 video file. I do not know the structure of the file. When I streamed the file via RTSP to VLC on my computer, I got roughly 20 seconds of video footage in full HD (1920x1080) followed by exactly the same footage repeated in SD. That is how I know that the first portion of `/tmp/view` is HD and the second portion of `/tmp/view` is SD.

**Comment by andy2301** [comment]
@xmflsct so, based on your experiments, this won't work if we upgrade to the 2.1.0.0A firmware? 

Anyway, I cleaned up the code a bit and attached here. Please check the README.txt file for instructions on compiling.
[rtsp2301_src.201704081250.zip](https://github.com/xmflsct/yi-hack-1080p/files/907771/rtsp2301_src.201704081250.zip)

Note that this is a different source package from the hp_ipc code. I couldn't make the hp_ipc one work.

Next step, let's take a look at how to stream the rmm processed video `/tmp/view`. @shadow-1 we need your help here. Do you mind sharing how you made it work streaming from /tmp/view? The rtsp2301 approach is a bit different: basically the sample_venc outputs the venc video feed into a shared memory segment and then the rtsp service picks it up and streams it.

**Comment by andy2301** [comment]
@shadow-1 the /tmp/view file doesn't seem to get updated (with a timestamp that's a couple of days old). VLC doesn't recognize the file as a video. Is this expected?

**Comment by xmflsct** [comment]
@andy2301 No, nor sample_venc or your RTSP implementation would work with 'new' firmware. Currently I am on the 'old' firmware and will track down which library/module that causes the problem. After then, I will move to 'new' firmware and try to compile again. I remember I was indeed using new libsns_f22.so for compiling. But I will try again to make sure. It is not only libsns_f22.so that causes problems. And I don't think f22 library will include in 050 SDK release anyway.

For me, VLC won't recognise /tmp/view, even if I change the extension to .264 or others. I try to stream /tmp/view using live555MediaServer on cam, but VLC on PC won't receive any signal.

**Comment by xmflsct** [comment]
Some updates.
1. Replacing libsns_f22.so will solve the problem of not supported fps.
2. Replacing hi3518e_isp.ko will solve the problem of config ISP register fail.
3. Replacing hi3518e_venc.ko is intended to solve the problem of HI_MPI_VENC_GetChnAttr gets stream fail. However, it seems that Linux hangs with kernel panic very soon after system starts booting. There is no auto-rebooting, and the LED stays blue of off. Now I will revert back to old firmware.

**Comment by xmflsct** [comment]
After several tries, my cam refuses to flash any (old or new) home firmware, but only rootfs.. No serial output to examine why. Finally, I try to place both old version of rootfs and home image on SD card and it works now.. Just that if you get stuck, this might help..

**Comment by shadow-1** [comment]
@andy2301 
I never checked the timestamp of the file. However whenever I used `vi` or `cat` on the file, it is clear that it is continuously updating. Perhaps the timestamp isn't being updated as the file is always open by `rmm` and never gets closed?

@xmflsct 
I did not use live555MediaServer on the camera. I compiled the live555 test programs and used them to test streaming `/tmp/view`. I used `testOnDemandRTSPServer` as it uses unicast transmission rather than multicast transmission. I could not get multicast transmission to work from the camera for some reason. I suspect multicast support has not been compiled into the kernel.

**Comment by shadow-1** [comment]
Actually I remember reading some time ago that multicast protocol is enabled in the kernel by default for quite a long time now. So it may not be an issue with the kernel. Perhaps it is an issue with the wifi chipset included with the camera or the driver used?

Either way, I could not get multicast to work at all no matter what I did and had to resort to RTSP streaming over unicast. However I could not get live555 to treat `/tmp/view` as a live streaming source. I could only manage to stream from it as if it were a prerecorded file.

**Comment by xmflsct** [comment]
@shadow-1 
Thank you so much! Yes, I can confirm that with `testOnDemandRTSPServer` I can indeed stream a copy of `/tmp/view`, and the file does get updated.

While running `rmm` manually, the program depends on `dispatch` to run before it does. And `dispatch` creates that `mmap.info` file, and `view` file is generated by `rmm`.
A side note. When I was working so hard to figure out the image sensor type, herewith the first three lines of output of `rmm`:
```
rmm.c(get_sensor_type-1573)[02:41:07.625]:f22 or f02
rmm.c(get_sensor_type-1594)[02:41:07.630]: 0xf
rmm.c(get_sensor_type-1603)[02:41:07.632]: 0x22
```

@andy2301 
Since I am on the 'old' home firmware now, I am trying to run your `rtsp2301` with `factory_test.sh`. However, there is no stream out to VLC somehow. If I run the program in telnet, everything is fine. I will see if I can have a workaround or understand better what is the problem.

**Comment by andy2301** [comment]
Is it possible that the watch_process kills the rtsp2301 or simply rebooted? 

Can you share the code that you use to steam /tmp/view? I'd like to save some time figuring that out.

**Comment by andy2301** [comment]
For the error "HI_MPI_VENC_GetChnAttr chn[0] failed with 0xffffffff", can you check the output of /dev/logmpp? Usually that log file has more info.

**Comment by shadow-1** [comment]
@andy2301 
Just need to download the live555 library and build the live555 test programs.
Source is available here: http://www.live555.com/liveMedia/public/

I modify `testOnDemandRTSPServer.cpp` and changed `Boolean reuseFirstSource = False;` to `Boolean reuseFirstSource = True;` in an attempt to get it streaming from the file continuously as per recommendation by the authors.

You can either change `char const* inputFileName = "test.264";` to `char const* inputFileName = "/tmp/view";` or create a symlink test.264 which points to /tmp/view.

**Comment by xmflsct** [comment]
@andy2301 I have playing for the whole afternoon, try to start `rtsp2301` from `factory_test.sh` but failed. I have no idea why. I use quite enough delay; change library path to fit with console one; etc. All I got is always:
```
size of 2RTSP server START
Streaming URL: rtsp://192.168.254.11:554/ch0_0.h264
listen for client connecting...
enSize=16, u32BlkSize=1491840
=============SAMPLE_COMM_VI_SetMipiAttr enWDRMode: 0
linear mode
sesnorID:0x80; sensoraddr:0x1; sensordata:0x1
=========================================================================
===F22sensor 1080P25fps(Parallel port) init success
=========================================================================
	 c) cbr.
please press twice ENTER to exit this sample
********************* Open pwm device error! *********************
[SAMPLE_COMM_VENC_GetVencStreamProc_RTSP]-75: HI_MPI_VENC_GetChnAttr chn[0] failed with 0xa0088005!
```
But when I run it from telnet console, or even wrap it in a script and run the script in telnet console, it works fine. Very weird. For above, I am referring to your version 1.

For version 2, I `cat /dev/logmpp` while run the program. Then I get 2 extra lines.
```
<3>[   sys] ioctl cmd does NOT exit!
<3>[   isp] [Func]:VREG_DRV_Exit [Line]:103 [Info]:<3>[   isp] The vreg of u32BaseAddr 0x10000 has not registerd!
```
Hope this is useful.

**Comment by andy2301** [comment]
Instead of using the factory_test.sh, did you try adding a simple file in  the /etc/init.d?

**Comment by andy2301** [comment]
The error in v1 seems to indicate that something was not initialized yet. Maybe not enough delay?

**Comment by xmflsct** [comment]
I just try to add it into `/etc/init.d/rcS` as a test, but the program won't even run. My guess is that, it is too early in the boot process, while those 3518 kernel modules are only loaded in `init.sh`. I also try to run it in `init.sh`, but no use... Very weird... I tried with a 5 minutes delay still no luck. I also try to telnet in as soon as the service starts up, and manually run the program and it is fine.

**Comment by andy2301** [comment]
another possibility between terminal and system script is the env settings. would it help to copy the related env settings into the automation script?

**Comment by xmflsct** [comment]
Yep, that's what I assumed as well. I notice there are some differences in both env settings. What I did was to specify library path and general path when I run `rtsp2301`, but that doesn't help. Are there more env settings that might be relevant?

**Comment by andy2301** [comment]
that's really weird. I can't think of other env setting but might worth a try to copy all env settings in script.

**Comment by xmflsct** [comment]
I will try in a bit, and let you know. :)

**Comment by xmflsct** [comment]
Changing all env settings won't help...

**Comment by andy2301** [comment]
ok. i think i see the problem now. It has some terminal interaction elements in the sample_venc part.

**Comment by xmflsct** [comment]
Nice! I will be waiting to test your next release. :)

(UPDATE)
Found [a way](http://stackoverflow.com/questions/1401002/trick-an-application-into-thinking-its-stdin-is-interactive-not-a-pipe) to trick the program and it works well!

**Comment by andy2301** [comment]
Try this one. I simply made the following changes in main.c. The previous one didn't run when put in background, but this one does.
[rtsp2301_20170409.zip](https://github.com/xmflsct/yi-hack-1080p/files/908602/rtsp2301_20170409.zip)


```
    // andy2301Mod: get rid of this interactive part to allow the program to run in background
    /*
    printf("please press twice ENTER to exit this sample\n");
    getchar();
    getchar();
    */

  while (!g_s32Quit) {
    sleep(1);
  }
```

**Comment by xmflsct** [comment]
This one cannot run in background yet, as in `sample_venc.c` there are two more parts that require key press - to select mode, and the same ENTER exit part.

**Comment by andy2301** [comment]
I don't think that source code is ever used. It's there just for reference.

**Comment by xmflsct** [comment]
You are right. I try to run it with `&` at the back, won't work. Put it into `factory_test.sh`, won't work neither.
But please don't be bothered by this minor issue yet. At least with [this trick](http://stackoverflow.com/questions/1401002/trick-an-application-into-thinking-its-stdin-is-interactive-not-a-pipe), it runs perfectly from script. I will push an update tomorrow with your program running in 'old' firmware with 'old' hack, and see if others would like to have a try and comment about it. :)

**Comment by andy2301** [comment]
Cool. I'm actually outside. Will try to check why it's not working in background tonight.

**Comment by andy2301** [comment]
@xmflsct the new version of rts2301 runs in background and I can stream from VLC. 
`/tmp/sd/tmp # ./rtsp2301 >& rtsp2301.txt &`

```
/tmp/sd/tmp # cat rtsp2301.txt 
Incoming RTSP connection accepted on socket: 16
EventLoop Connection reached: 3
192.168.101.139 Port:63855
S192.168.101.139 Port:63855
S192.168.101.139 Port:63855
*** transport: RTP/AVP;unicast;client_port=54926-54927 ***
TRANSFER TO READY STATE!
S192.168.101.139 Port:63855
        Start Playing!
S192.168.101.139 Port:63855
Sfd:16,RTSP connection closed by client.
```

**Comment by xmflsct** [comment]
@andy2301 I tried again, and it does run in background. Maybe I copy the wrong binary yesterday, late at night, I am so sorry.

**Comment by maysamsh** [comment]
Do you have any estimate of release a working firmware?

**Comment by xmflsct** [comment]
@maysamsh I am working on a final testing now to make sure everything is working. I should push an update in half an hour.

**Comment by maysamsh** [comment]
@xmflsct I'm excited to test it.

**Comment by xmflsct** [comment]
@maysamsh I just push an update. Please read carefully the readme document as there are some requirements before using this hack.

**Comment by maysamsh** [comment]
Okay. thanks.

**Comment by maysamsh** [comment]
@xmflsct It does not accept root password. I have not changed it in the `yi-hack-1080p.cfg` file. The LED status is still blue.

**Comment by xmflsct** [comment]
@maysamsh Can you create a new issue, and let's discuss there? I would like to dedicate this thread to only about further RTSP development with @andy2301.

**Comment by maysamsh** [comment]
@xmflsct sure, the password was blank! now I have access to the terminal, but the official app does not work, and I don't have access to the RTSP stream.

**Comment by xmflsct** [comment]
@andy2301 Somehow, I could not even get my sample_venc to work again somehow. :( Same error as I mentioned above about ISP error, and I could not trace down the problem.

By the way, two findings:
1. [hi35xx/live-streamer](https://github.com/hi35xx/live-streamer), this seems to be interesting?
2. While I `cat /proc/media-mem` with `rmm` running, this seems to be the mapping of all video buffers in the memory. Maybe this can be directly feed into RTSP stream?

**Comment by shadow-1** [comment]
@xmflsct 
Interesting find with the live-streamer. I won't be able to look into it for a while. Shame that the author decided to write his project in C++ and not straight C if it turns out to be a promising library.

**Comment by xmflsct** [comment]
@shadow-1 
No worries. I think we are progressing very fast already. :)
Sadly my limited knowledge to C/C++ cannot contribute that much of you guys' programming. But I am up for any testing/debugging/researching/etc./etc., to save your time!

**Comment by vaidyasr** [comment]
@andy2301 
Link for 2.0.0.1A_201612051401 
http://download.xiaoyi.com/smarthomecam/2.0.0.1A_201612051401home_y20m

**Comment by andy2301** [comment]
guys, i cooked another version of rtsp server that streams from /tmp/view continuously. This one should work side by side with the existing stock features. So far it's only tested on the original 2.0.0.1A firmware.
[rtsp2303_live_bin_20170414-1630.zip](https://github.com/xmflsct/yi-hack-1080p/files/923990/rtsp2303_live_bin_20170414-1630.zip)

**Comment by andy2301** [comment]
Remember that my prior streaming solution rtsp2301 requires to shut down the stock streaming process `rmm`, rendering the YiHome app function and all other motion detection functionalities unavailable. So, I started looking for a different solution that co-exists with the stock features.

Thanks to @shadow-1, who mentioned that the `rmm` process writes two streams 1080p and 360p to file /tmp/view with 10 seconds of video data. So this rtsp2303 solution focuses on streaming from /tmp/view file (1586752 bytes).

Think of the /tmp/view as a loop file that the `rmm` process writes to continuously (start over once it reaches the end-of-file offset 1586752. So, in order to stream from this file, we need to solve the following problems:
1. **find out the file offset boundary of the 1080p stream.** 
I've experimented a lot and it seems to me that offset range [0, 1048576] (the first 1MB) is a safe range for 1080p stream.  Sometimes [0, 1152000] gives you more frames of 1080p video, but other times it gives you the 360p video. So in this program I sticked to the safe range of [0, 1048576] or [0, 1050000].
2. **find out which part of the offset range [0, 1048576] to start, in order to reduce the lagging issue.** 
If you start from offset 0, your lag is random, and could range from [1-10] seconds. I did implement a function "monitorChanges" to detect (at the beginning of streaming or could be set to run periodically) the latest changes and start streaming from there. 
It works to reduce the lag most of time to [2-4] seconds. But I couldn't get it better than that. Once you get too close, you will get garbled pictures, probably because the writing of latest 1080p data is still in flux to file. 
Anyhow, you are welcome to try it out by yourself with the env variable settings (adjIDX), as demonstrated in the rtsp2303.sh script.

In summary:
**Pros:**
- Co-exist with all the stock features
- Much less overhead. The cpu usage is only 12% while streaming 1080p at 15fps (the same frame rate written to /tmp/view and the recorded .mp4 files).

**Cons:**
- The streamed video lags behind [2-4] seconds most of time, but sometimes it could lag up to 10 seconds.
- It is not stable. After some time (30 seconds ~ 4 minutes observed from my experiments) the streaming client app may stop. The reason, I suspect, is that the PCR timestamp in the /tmp/view confuses the client app and it simply can't handle any more after some time.

Let me know if anybody has better approaches to solve the 2 problems:
- the exact offset range of the 1080p stream in /tmp/view
- where to start reading from the file at any given time, in order to reduce the lagging

**Comment by andy2301** [comment]
There are some of the variables used in rtsp2303. You can change these default values from the command line by setting env variables (as demonstrated in rtsp2303.sh script).
```
unsigned long startOFS = 0;             // the starting file offset to read from /tmp/view
unsigned long repeatOFS = 1050000;   // we rewind the file to the starting file offset when it reaches this repeatOFS. Basically we only read [start, repeat].
unsigned int numCHUNK = 7;             // we split the [start, repeat] offset range of /tmp/view into numCHUNK chunks, in order to determine which chunk has changed
unsigned long sleepMS = 500;          // read a buf from each chunk of /tmp/view, sleep for sleepMS, and read another buf to compare which chunk changed
unsigned long cmpSZ = 10240;           // the number of bytes of data to read from each chunk of /tmp/view
int adjIDX = -1;                                    // if we detect that chunk[5] changed, we may want to start reading from chunk[5 + adjustIdx]
int readsPMON = 0;                   // reads per monitor: how many doReadFromFile invokes (file reads) before we run each monitorChanges.  0 = disable monitorChanges
int VERB = 0;                             // print out more info if set to 1
```

The following is the logic of the `monitorChanges` implementation.
```
// Monitor which part of file /tmp/view just got changed, so that we can read from there to get the latest video stream.
// The purpose is to make sure we stream the latest video segments rather than possibly 10 seconds lagging behind (file /tmp/view stores roughly 10 sec of video)
// 1. divide file offset range [startOFS, repeatOFS] of /tmp/view into num_chunks chunks (logically)
// 2. take a snapshot of <cmpSZ> bytes of data from each chunk [chunkStartOfs, chunkStartOfs+cmpSZ]
// 3. sleep for <sleepMS> mili-seconds
// 4. take another snapshot of <cmpSZ> bytes of data from each chunk [chunkStartOfs, chunkStartOfs+cmpSZ]
// 5. for each chunk, compare the two snapshots for changes. Mark the chunk as changed if any.
// 6. return the chunkStartOfs offset of the chunk that just got changed if adjIDX=0, otherwise, return the offset of chunk id (changedChunkIdx + adjIDX).
//    If adjIDX = -1, return the offset of the prior chunk. This is the default. The rationale is if we read the current changing chunk, we might get fragmented frame/data.
//    If adjIDX = -2, return the offset of the prior chunk. This is safer to avoid fragmented frame/data but may have increased lagging.
//    If adjIDX = 1, return the offset of the next chunk.
//    .....
//
```

**Comment by shadow-1** [comment]
@andy2301 
Great work! I won't have a lot of time to test this yet.
However I do have an observation:

Are you sure `/tmp/view` is a circular buffer? From my experimentation with the live555 library in the past, I never seen behaviour that suggests that it could be a circular buffer. From my experimentation, I was led to believe that it is probably a simple FIFO buffer, so the latest video data is located in the same area in the file. So the video that is roughly 10 seconds behind live feed is always towards the start of the file whilst the closest to 'live' video is towards the end of the stream in `/tmp/view`.

Unfortunately I do not know the offset between the 1080p and the 360p streams.

**Comment by andy2301** [comment]
@shadow-1 I think it's a circular buffer.
Following your post, I ran the following to always stream the 150k bytes [900000, 1050000]. It simply repeats the 1~2 seconds of video frames, for example 18:12:41 - 42, and then after 10 seconds it jumps forward to 18:12:50 - 51.

`startOFS=900000 repeatOFS=1050000 VERB=1 readsPMON=-1 ./rtsp2303`

**Comment by shadow-1** [comment]
@andy2301 
Ah ok, no problems.

With regards to the offsets, I have a suspicion that it will be either located somewhere within `/tmp/mmap.info` or be hardcoded within the application.

When I get more free time, I can disassemble the Xiaomi applications and trace what is being read from `/tmp/mmap.info`.

**Comment by andy2301** [comment]
Please do when you get time. Thanks. @shadow-1

**Comment by xmflsct** [comment]
@andy2301
Wonderful progress! I am away currently, and will be back for helping testing in three days.

**Comment by xmflsct** [comment]
@andy2301 I am back, and just test a bit. Indeed after every reboot, `rtsp2303` performs differently as per chunk data varies. `mmap.info` also varies a bit after each reboot, but I am not sure if those differences represent chunk size or offset. But at least in the v1 example, `mmap.info` is used somehow.

To provide a separate idea that might be useful. Since when streaming with live555 of `/tmp/view`, it performs rather stable of each connection. Maybe, we could use h264 headers to determine each chunk/frame to adapt to the file dynamically? I will look into if I can analyse `/tmp/view` from h264 header perspective and update if I find something useful.

**Comment by andy2301** [comment]
Sorry for late reply. I went on a trip and just got back.

I didn't get much useful info from mmap.info. Maybe @shadow-1 can help us there.

@xmflsct yes it'll be great if you can figure out how to read the h264 headers and adapt to the file dynamically. rtsp2303 is based on live555. I'm attaching the source code. My modification involves mainly `rtsp2303.cpp` and `livemedia/ByteStreamFileSource.cpp`. You can do a global search of "andy2301Mod" for other changes (e.g. FrameRates changed from 25 to 15).
[rtsp2303_srcbin_20170414-1630.zip](https://github.com/xmflsct/yi-hack-1080p/files/965411/rtsp2303_srcbin_20170414-1630.zip)

**Comment by xmflsct** [comment]
@andy2301 No worries. Hope you have had a great time. :)

Recently I was also thinking about using live555 as a base, since the test program can stream /tmp/view without problem, and we just have to somehow tell it to restart every `x` seconds. I will take a look into your source code.

**Comment by shadow-1** [comment]
@andy2301 @xmflsct 
I hope to find some time to start working on RTSP soon!

I like Live555 because it is a very mature library that is actually used in many commercial applications. However what I don't like about it is that it is written in C++ which makes the file size quite large (in comparison with C programs) and of course there is a dependency on the standard C++ library libstdc++.

@xmflsct 
I think you are right, I think we should be working out where to start streaming from and then continue streaming from there on whilst rolling back to the start of the file when it reaches an offset. Easier said than done of course.

@andy2301 
Did you compile your own version of libstdc++?

What really put me off C++ applications on these cameras is that libstdc++ that comes with the SDK is almost 5mb in size! That is absolutely ridiculously large which makes C++ programs completely unsuitable for these cameras as they have 16mb of flash memory in total.

However I noticed that libstdc++ that you included in your source file is under 1mb in size. Much more manageable.

**Comment by xmflsct** [comment]
@shadow-1 
That's wonderful! I will rely on you guys to take RTSP service further. :)

**Comment by andy2301** [comment]
The libstdc++ lib is from the Yi 720p v1 firmware. I used a stripper called upx which reduced the rtsp2303 footprint to less tha 200k.

**Comment by xmflsct** [comment]
@andy2301 I have put my 1080p camera into production together with my other original 720p one. It runs now with your `rtsp2301` process. It sometimes crashes so I add a watch process to restart it when this happens. Luckily with current setup, I can easily update files remotely and test new setups. Please let me know if you have further progress with `rtsp2303`. :)

**Comment by naterh** [comment]
@shadow-1 late to the party but, looks like there are some http streaming add-ons included with the SDK 

http://blog.51cto.com/shugenyin/1851241

**Comment by saundersrg** [comment]
You people are legends. I've got RTSP running on a Yi Dome 1080p with @shadow-1 firmware and @andy2301 script, thank you!

**Comment by gelsas** [comment]
@saundersrg could you share the final files you used and a short tutorial? I would love to use it aswell on my dome 1080p

**Comment by saundersrg** [comment]
It really didn't work as well as I'd hoped tbh, and was really simple. I've ended up selling the Yi and buying a Foscam FI9821P. Comparable pricing with full ONVIF, RTSP and support in Zoneminder.

Firstly I flashed with the firmware from @shadow-1, and made sure I could SSH. I then took the scripts from @andy2301 moved them (WinSCP/PuTTY) into a folder under /home/ and tested that I could launch manually. The feed would work for about 10s and then the client device would disconnect (be it VLC or ZM), and I'd have to keep enabling and disabling the monitor in ZM. For some reason I couldn't get Andy's script to launch automatically at any stage, using rc or the init scripts.

**Comment by andy2301** [comment]
@saundersrg the client device disconnecting issue is expected, as I mentioned in my previous comment when I published `rtsp2303`. Recently I got some help from another thread. Might be able to fix this if things go well.

**Comment by squirrel289** [comment]
Hey @andy2301!  Thanks for sharing your work with everyone.  It seems that, between your contributions and those of the several base images (like this one and @shadow-1) many have _nearly_ managed to get a working rtsp solution on these puppies.  I wanted to see if you'd managed to incorporate the latest changes you referred to in your last note.  If so, any chance you'd be willing to share?  If not, would you be willing to point others to the same advice and to publish your current work to a public repo so others can fork and/or create pull requests?

Thanks,
BS
